File: netlify/functions/collaborate.js
// Converted from collaborate.py
// Netlify Function: Handle collaboration submissions
// Saves to netlify/data/collaborations.json

const fs = require('fs');
const path = require('path');

// Map user input regions to the three defined regions
const REGION_MAPPING = {
  eastern: ['eastern', 'east', 'makran', 'gwadar', 'balochistan east', 'coast', 'coastal'],
  western: ['western', 'west', 'kalat', 'zhob', 'balochistan west', 'highland', 'mountain'],
  southern: ['southern', 'south', 'rakhshan', 'desert', 'balochistan south', 'plain']
};

function mapRegionToId(userRegion) {
  /**Map user input region to one of the defined region IDs.*/
  if (!userRegion) {
    return null;
  }
  
  const userRegionLower = userRegion.toLowerCase().trim();
  
  for (const [regionId, keywords] of Object.entries(REGION_MAPPING)) {
    for (const keyword of keywords) {
      if (userRegionLower.includes(keyword)) {
        return regionId;
      }
    }
  }
  
  return null;
}

exports.handler = async (event, context) => {
  // Handle collaboration POST requests
  
  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  }
  
  try {
    // Parse body data
    let data = {};
    
    if (event.body) {
      if (event.body.startsWith('{')) {
        data = JSON.parse(event.body);
      } else {
        // Simple form parser
        const pairs = event.body.split('&');
        for (const pair of pairs) {
          if (pair.includes('=')) {
            const [k, v] = pair.split('=');
            data[k] = decodeURIComponent(v);
          }
        }
      }
    }
    
    const name = (data.name || '').trim();
    const category = (data.category || '').trim();
    const title = (data.title || '').trim();
    const regionInput = (data.region || '').trim();
    const content = (data.content || '').trim();
    
    if (!name || !category || !title || !content) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: 'All fields required' })
      };
    }
    
    // Map user region input to a defined region ID
    const regionId = mapRegionToId(regionInput);
    
    // Create collaboration entry
    const entry = {
      name,
      category,
      title,
      region: regionInput,
      region_id: regionId,
      content,
      filename: null,
      timestamp: new Date().toISOString().replace('T', ' ').split('.')[0] + ' UTC'
    };
    
    // Save to file
    const dataDir = path.join(__dirname, '../../dist/data');
    if (!fs.existsSync(dataDir)) {
      fs.mkdirSync(dataDir, { recursive: true });
    }
    
    const collabFile = path.join(dataDir, 'collaborations.json');
    
    // Load existing collaborations
    let collabs = [];
    if (fs.existsSync(collabFile)) {
      const fileContent = fs.readFileSync(collabFile, 'utf-8');
      collabs = JSON.parse(fileContent);
    }
    
    // Add new collaboration
    collabs.unshift(entry);
    
    // Save updated collaborations
    fs.writeFileSync(collabFile, JSON.stringify(collabs, null, 2), 'utf-8');
    
    return {
      statusCode: 200,
      body: JSON.stringify({ success: true, message: 'Collaboration saved!' })
    };
  } catch (error) {
    console.error('Error:', error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: 'Internal server error' })
    };
  }
};

File: netlify/functions/collaborate.py
"""
Netlify Function: Handle collaboration submissions
Saves to dist/data/collaborations.json
"""

import json
import os
from datetime import datetime
from pathlib import Path

# Map user input regions to the three defined regions
REGION_MAPPING = {
    'eastern': ['eastern', 'east', 'makran', 'gwadar', 'balochistan east', 'coast', 'coastal'],
    'western': ['western', 'west', 'kalat', 'zhob', 'balochistan west', 'highland', 'mountain'],
    'southern': ['southern', 'south', 'rakhshan', 'desert', 'balochistan south', 'plain']
}

def map_region_to_id(user_region):
    """Map user input region to one of the defined region IDs."""
    if not user_region:
        return None
    user_region_lower = user_region.lower().strip()
    for region_id, keywords in REGION_MAPPING.items():
        for keyword in keywords:
            if keyword in user_region_lower:
                return region_id
    return None

def handler(event, context):
    """Handle collaboration POST requests"""
    
    if event['httpMethod'] != 'POST':
        return {
            'statusCode': 405,
            'body': json.dumps({'error': 'Method not allowed'})
        }
    
    try:
        # Parse form data or JSON
        body = event.get('body', '{}')
        
        # Handle both JSON and form-encoded data
        if isinstance(body, str):
            if body.startswith('{'):
                data = json.loads(body)
            else:
                # Simple form parser
                data = {}
                for pair in body.split('&'):
                    if '=' in pair:
                        k, v = pair.split('=', 1)
                        data[k] = v
        else:
            data = body
        
        name = data.get('name', '').strip()
        category = data.get('category', '').strip()
        title = data.get('title', '').strip()
        region_input = data.get('region', '').strip()
        content = data.get('content', '').strip()
        
        if not all([name, category, title, content]):
            return {
                'statusCode': 400,
                'body': json.dumps({'error': 'All fields required'})
            }
        
        # Map user region input to a defined region ID
        region_id = map_region_to_id(region_input)
        
        # Create collaboration entry
        entry = {
            'name': name,
            'category': category,
            'title': title,
            'region': region_input,
            'region_id': region_id,
            'content': content,
            'filename': None,
            'timestamp': datetime.utcnow().strftime('%Y-%m-%d %H:%M UTC')
        }
        
        # Save to file
        data_dir = os.path.join(os.path.dirname(__file__), '../../dist/data')
        os.makedirs(data_dir, exist_ok=True)
        
        collab_file = os.path.join(data_dir, 'collaborations.json')
        
        # Load existing collaborations
        collabs = []
        if os.path.exists(collab_file):
            with open(collab_file, 'r', encoding='utf-8') as f:
                collabs = json.load(f)
        
        # Add new collaboration
        collabs.insert(0, entry)
        
        # Save updated collaborations
        with open(collab_file, 'w', encoding='utf-8') as f:
            json.dump(collabs, f, ensure_ascii=False, indent=2)
        
        return {
            'statusCode': 200,
            'body': json.dumps({'success': True, 'message': 'Collaboration submitted!'})
        }
    
    except Exception as e:
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }
